man/man2/select.2
.\" Hey Emacs! This file is -*- nroff -*- source.
.\"
.\" This manpage is copyright (C) 1992 Drew Eckhardt,
.\"                 copyright (C) 1995 Michael Shields.
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a
.\" permission notice identical to this one
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not
.\" have taken the same level of care in the production of this manual,
.\" which is licensed free of charge, as they might when working
.\" professionally.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by
.\" the source, must acknowledge the copyright and authors of this work.
.\"
.\" Modified 1993-07-24 by Rik Faith <faith@cs.unc.edu>
.\" Modified 1995-05-18 by Jim Van Zandt <jrv@vanzandt.mv.com>
.\" Sun Feb 11 14:07:00 MET 1996  Martin Schulze  <joey@linux.de>
.\"	* layout slightly modified
.\"
.\" Modified Mon Oct 21 23:05:29 EDT 1996 by Eric S. Raymond <esr@thyrsus.com>
.TH SELECT 2 "February 11, 1996" "Linux 1.2" "Linux Programmer's Manual"
.SH NAME
select, FD_CLR, FD_ISSET, FD_SET, FD_ZERO \- synchronous I/O multiplexing
.SH SYNOPSIS
.B #include <sys/time.h>
.br
.B #include <sys/types.h>
.br
.B #include <unistd.h>
.sp
\fBint select(int \fIn\fB, fd_set *\fIreadfds\fB,
fd_set *\fIwritefds\fB, fd_set *\fIexceptfds\fB,
struct timeval *\fItimeout\fB);
.sp
.BI "FD_CLR(int " fd ", fd_set *" set );
.br
.BI "FD_ISSET(int " fd ", fd_set *" set );
.br
.BI "FD_SET(int " fd ", fd_set *" set );
.br
.BI "FD_ZERO(fd_set *" set );
.fi
.SH DESCRIPTION
.B select
waits for a number of file descriptors to change status.
.PP
Three independent sets of descriptors are watched.  Those listed in
.I readfds
will be watched to see if characters become
available for reading, those in
.I writefds
will be watched to see if it is ok to immediately write on them, and
those in
.I exceptfds
will be watched for exceptions.  On exit, the sets are modified in place
to indicate which descriptors actually changed status.
.PP
Four macros are provided to manipulate the sets.
.B FD_ZERO
will clear a set.
.B FD_SET
and
.B FD_CLR
add or remove a given descriptor from a set.
.B FD_ISSET
tests to see if a descriptor is part of the set; this is useful after
.B select
returns.
.PP
.I n
is the highest-numbered descriptor in any of the three sets, plus 1.
.PP
.I timeout
is an upper bound on the amount of time elapsed before
.B select
returns.  It may be zero, causing
.B select
to return immediately.  If
.I timeout
is NULL (no timeout),
.B select
can block indefinitely.
.SH RETURN VALUE
On success,
.B select 
returns the number of descriptors contained in the descriptor sets, which
may be zero if the timeout expires before anything interesting happens.
On error, \-1 is returned, and
.I errno
is set appropriately; the sets and
.I timeout
become undefined, so do not
rely on their contents after an error.
.SH ERRORS
.TP 0.8i
.B EBADF
An invalid file descriptor was given in one of the sets.
.TP
.B EINTR 
A non blocked signal was caught.
.TP
.B EINVAL 
.I n
is negative.
.TP
.B ENOMEM
.B select
was unable to allocate memory for internal tables.
.SH NOTES
Some code calls
.B select
with all three sets empty,
.I n
zero, and a non-null
.I timeout
as a fairly portable way to sleep with subsecond precision.
.PP
On Linux,
.I timeout
is modified to reflect the amount of time not slept; most other
implementations do not do this.  This causes problems both when Linux
code which reads
.I timeout
is ported to other operating systems, and when code is ported to Linux
that reuses a struct timeval for multiple
.BR select s
in a loop without reinitializing it.  Consider
.I timeout
to be undefined after
.B select
returns.
.SH EXAMPLE
.nf
#include <stdio.h>
#include <sys/time.h>
#include <sys/types.h>
#include <unistd.h>

int
main(void)
{
    fd_set rfds;
    struct timeval tv;
    int retval;

    /* Watch stdin (fd 0) to see when it has input. */
    FD_ZERO(&rfds);
    FD_SET(0, &rfds);
    /* Wait up to five seconds. */
    tv.tv_sec = 5;
    tv.tv_usec = 0;

    retval = select(1, &rfds, NULL, NULL, &tv);
    /* Don't rely on the value of tv now! */

    if (retval)
        printf("Data is available now.\\n");
        /* FD_ISSET(0, &rfds) will be true. */
    else
        printf("No data within five seconds.\\n");

    exit(0);
}
.fi
.SH CONFORMING TO
4.4BSD (the
.B select
function first appeared in 4.2BSD).  Generally portable to/from
non-BSD systems supporting clones of the BSD socket layer (including
System V variants).  However, note that the System V variant typically
sets the timeout variable before exit, but the BSD variant does not.
.SH SEE ALSO
.BR accept (2),
.BR connect (2),
.BR read (2),
.BR recv (2),
.BR send (2),
.BR write (2).


changes.txt
File: lsck.inf,  Node: Changelog,  Next: License,  Prev: Credits,  Up: Top

Changelog
*********

Version 0.8.1 ????
==================

These changes are in no particular order:

   * `recv()' and `send()' should not hang the DOS box anymore, when
     waiting for input/output with SOCK.VXD and Winsock 2.

   * Various texinfo documentation updates, to be consistent with DJGPP
     libc's documentation style.

Version 0.8.0 2000-11-16
========================

These changes are in no particular order:

   * Fixed bug in version string construction (*note
     __lsck_get_version::).

   * Documentation fixes:

        * Clarified `FIONBIO' versus `O_NONBLOCK' (*note ioctl_list::).

        * `gethostname()' page now describes how the host name is
          discovered (*note gethostname::).

   * Fixed the DSMs for installation with pakke.

   * Fixed a bug in the auto-configuration code, where some network
     cards were ignored due to a bug in the registry entry parsing.

   * Added `tags', `id' and `ID' targets to Makefile, for building
     Emacs tags and GNU id-utils files respectively.

   * Fixed a big in `gethostname()'; previously EINVAL was erroneously
     returned for buffers larger than `MAXHOSTNAME'. This broke the name
     resolving code for automatic configuration.

   * Documented how bash and libsocket's auto-configuration code
     interacts with regard to `$HOSTNAME' (*note libsocket and
     environment variables::).

   * Fixed the DNS server questions in netsetup (*note Introduction:
     (netsetup)Introduction.). The input routines also now perform more
     validation.

   * Fixed header files so that #include lines are before checks for C++
     (extern "C", etc.).

   * Fixed header `netinet/in.h' to work around a compiler bug in g++
     2.95.x.  The structure `struct ip_opts' is not defined, when using
     g++ 2.95.x.

   * "Beej's Guide to Network Programming" and the libsocket demos are
     now included in the binary distribution.

Version 0.8.0 Pre 1 2000-09-11
==============================

These changes are in no particular order:

   * Texinfo documentation: corrected copyright message to include
     Alain Magloire in info version; now distributed under GNU FDL.

   * Fixed assembly code to work with binutils 2.9.5.1 beta and later.
     Hopefully libsocket will build with _all_ versions of binutils
     now.  Thanks to Eli Zaretskii for his help & suggestions.

   * Problems using libsocket with firewall software were discovered
     and documented (*note Known Bugs::). Thanks to Robert, Tim van
     Holder.

   * The demo `solist' now builds on Linux.

   * The source distribution requires GNU shellutils to build; this is
     now documented.

   * The binary distribution had dependency generation for the demo and
     netsetup directories.  This is no longer included - it was not
     very useful and it meant that GNU shellutils was required to build
     (only `echo' was actually required).

   * `all' target now fixed in source distribution, so that it doesn't
     run `netsetup' by default.  This problem stopped cross-compilation
     from completing.

   * The error code `ELOOP' is now defined conditionally in
     `include/lsck/errno.h', so as not to conflict with DJGPP 2.04's
     symlink support (DJGPP 2.04 is currently in development).  Thanks
     to Laurynas Biveinis.

   * Finished error messages for all libsocket-specific error messages
     in `src/newerror.c'.

   * Fixed the code for parsing the DHCP settings from the registry.
     Thanks to Tim van Holder.

Version 0.7.4 Beta 4 2000-05-28
===============================

These changes are in no particular order:

   * Added a guide to configuration and usage of libsocket programs (a
     user guide), to be distributed with programs built using
     libsocket.  This was added to the `redist' directory.

   * The auto-configuration from the registry was found to be buggy and
     was fixed.

   * Problems were found compiling with gcc 2.95.x:

        * The inline assembly used by several of the source files was
          buggy, but was compiled without error with previous versions
          of gcc.  The assembly code was fixed.  See the section in the
          gcc FAQ about spilled registers.

        * The resolver code used pragmas to generate weak symbols that
          could be overridden by Linux threading libraries.  These
          generated warnings with gcc 2.95.2 (whose DJGPP port had no
          weak symbol support at the time) and errors with a
          cross-compiler (egcs-2.91.66 19990314 (egcs-1.1.2 release))
          on Linux.  The pragmas were removed.  This involved
          implementing `readv()' and `writev()'.


   * A new demo was added, `solist', to list the available socket
     options for UDP and TCP options.

   * netsetup previously only configured one DNS server - now the user
     can specify multiple DNS servers.

   * The top-level source Makefile, `Makefile.src', used to create the
     directory `lib' when building the library.  This was moved to
     `src/Makefile', since this is where the library is actually built.

   * The `uninstall' Makefile target previously failed to uninstall some
     files from `$DJDIR/include/arpa'.  This has been fixed.

   * The `socketpair()' function has been implemented, but only for
     Unix domain sockets (see below).

   * The Unix domain socket support is now much better.  It is still
     alpha-quality code, which means there are definitely some
     bugs/problems with it.

   * Added the POSIX function `sockatmark()', which always fails.

   * Changes were made to better conform with Unix98 description of the
     sockets interface: added `SHUT_*' definitions for `shutdown()';
     use of `size_t', `ssize_t' instead of `int'.

   * Fixed `fcntl()' so that it accepts more than one option
     simultaneously.  Previously passing more than one option OR'd

demo/select.c
/*
 * Demo programme for libsocket for DJGPP <C> 1998 Indrek Mandre.
 * Sample programme shows how to use select. This is C language specific.
 * That means it is described in all C books.
 * Select lets you to check if there's anything to read from some
 * file descriptor or wait until something arrives or wait some time
 * something to arrive. In libsocket there's no difference between
 * socket descriptor and file descriptor. You can use the same functions on
 * both of them (eg. select). To test if it's right to do accept use the same
 * way. And please read documentation of djgpp. It's explained there more
 * closely.
 * This file will compile and work also on Linux.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>     /* RD: Needed for memset */
#include <sys/types.h>
#include <sys/time.h>
#include <unistd.h>

int main()
{
 fd_set fd;
 struct timeval tv;
 char str[256];

 printf ("Demo program for libsocket. Demonstrates the basic use of select.\n");
 printf ("It waits from user input and prints dots at the same time.\n");
 printf ("Here we go...\n");

 for (;;)
 {
     FD_ZERO ( &fd ); /* We make all the bits to zero */
     FD_SET ( fileno ( stdin ), &fd ); /* We want to check if there's something to read */

 tv.tv_usec = 0;
 tv.tv_sec = 1;     /* We wait 1 seconds for the data. If all are zeros then
                       we wait 0 seconds and 0 milliseconds. If we pass
                       NULL to select as tv, then we'll wait for ever */

     select ( fileno (stdin) + 1, &fd, 0, 0, &tv );

     if ( FD_ISSET ( fileno ( stdin ), &fd ) )
     {
        gets ( str );
        printf ( "Got it: [%s]\n", str );
     }

     printf (".");
     fflush ( stdout );
 }

}

